enhancements/ignition-spec-dual-support.md

```diff
@@ -0,0 +1,447 @@
+- [ ] Create tests to apply spec 3 configs to spec 2 clusters successfully
+- [ ] An alerting mechanism is put in place for outdated and incompatible/non-translatable configs
+- [ ] Docs are updated to reflect the new config version
+
+#### Phase 1 (optional):
+- [ ] MCC attempts to translate all machineconfigs with v2 ignition config to v3, leaving those which cannot be converted and warning the user of future changes
+- [ ] MCC learns to generate v3 initial master/worker ignition configs
+
+#### Phase 2:
+- [ ] The MCO gains the ability to manage installer-generated stub master/worker ignition configs (separate enhancement proposal)
+- [ ] ignition-dracut spec 2 and spec 3 diffs are aligned
+- [ ] RHCOS bootimages switches to only accept ignition spec 3 configs
+- [ ] The Openshift installer is updated to generate spec 3 configs
+- [ ] The MCC gains the ability for users to provide necessary changes to update spec 2 to spec 3
+- [ ] MCO enforces that all configs are spec 3 before allowing the CVO to start the update
+- [ ] Further tests/docs are added
+
+
+### Non-Goals
+
+ - Support FCoS/OKD directly through this change
+ - API support for MCO, namely switching to RawExt formatted machineconfig
+ objects instead of explicitly referencing ignition, is not considered as
+ part of this proposal
+
+
+## Proposal
+
+This change is multi-component:
+
+
+### Vendoring Changes
+
+The MCO and Installer must change to go modules (currently dep) for vendoring
+as ignition v2 requires using go modules. To support both spec 2 and spec 3,
+both ignition versions must be vendored in for typing. 
+
+
+### Spec Translator
+
+To handle both spec versions, as we need the ability to upgrade existing
+clusters, we will create a translator between spec 2 and spec 3. This ensures
+that a cluster only has one “desiredConfig” which will be translated to spec 3
+when the MCO with dual support detects that the existing configuration of a
+machine is on spec 2 (will happen only once for all existing and new nodes,
+when the MCO with dual support is first deployed onto the cluster). This will
+only be required as part of the MCO.
+
+Note that there exists three types of spec 2 configs:
+ - Those that are directly translatable to spec 3. This is the case for all existing IPI configs.
+ - Those that we’re not 100% sure we can directly translate, but we can infer what the user is doing and do a translation 
+ - Those that we’re 100% sure we CANNOT translate directly, and requires user input for us to correctly translate
+
+During phase 1 we should attempt to translate on a best-effort basis. If the
+cluster can be directly translated to spec 3, we will do so and use the now
+v3 spec version config in the MCD. If not, we will support both, and alert
+the user that there are untranslatable configs.
+
+During phase two, we should fail updates unless the cluster is fully on spec
+3 config. This effectively means that UPI clusters are at risk when upgrading
```

Comment:
Why are UPI clusters at risk?

Reply:
More likely the user supplied their own ignition config customizations